File: strongly_connected_components/go/adjacency_list.go
package strongly_connected_components

type Graph struct {
	VertexCount int
	Directed    bool
	Edges       []Edge
}

type Edge struct {
	X int
	Y int
}

type AdjacencyList [][]int

type VertexCallback func(x int)

func NewAdjacencyList(g Graph) AdjacencyList {
	a := make(AdjacencyList, g.VertexCount)
	for _, edge := range g.Edges {
		insertEdge(a, edge.X, edge.Y, g.Directed)
	}
	return a
}

func insertEdge(a AdjacencyList, x, y int, directed bool) {
	a[x] = append(a[x], y)
	if !directed {
		insertEdge(a, y, x, true)
	}
}

func (a AdjacencyList) DepthFirstSearch(start int, late VertexCallback, discovered []bool) {
	if discovered[start] {
		return
	}
	discovered[start] = true
	for _, neighbor := range a[start] {
		a.DepthFirstSearch(neighbor, late, discovered)
	}
	late(start)
}

File: topological_sort/typescript/adjacency_list.ts
export interface Graph {
    vertexCount: number;
    directed: boolean;
    edges: Edge[];
}

export interface Edge {
    x: number;
    y: number;
    weight: number;
}

type VertexCallback = (x: number) => void;

export class AdjacencyList {
    a: Edge[][];

    constructor(graph: Graph) {
        this.a = constructAdjacencyList(graph);
    }

    depthFirstSearch(start: number, lateCallback: VertexCallback, discovered: boolean[]) {
        if (discovered[start]) {
            return;
        }

        discovered[start] = true;

        for (const edge of this.a[start]) {
            this.depthFirstSearch(edge.y, lateCallback, discovered);
        }

        lateCallback(start);
    }
}

function constructAdjacencyList(graph: Graph): Edge[][] {
    // new Array(vertexCount).fill([]) does not work because it reuses the same array instance for every element.
    const a = Array.from({length: graph.vertexCount}, () => []);

    for (const edge of graph.edges) {
        insertEdge(a, edge, graph.directed);
    }

    return a;
}

function insertEdge(a: Edge[][], edge: Edge, directed: boolean): void {
    a[edge.x].push(edge);
    if (!directed) {
        const reversed = {x: edge.y, y: edge.x, weight: edge.weight};
        insertEdge(a, reversed, true);
    }
}
